chrome/content/zotero/xpcom/itemTreeManager.js

```diff
@@ -0,0 +1,370 @@
+		}
+		return "";
+	}
+
+	/**
+	 * Get columns by type. Only support itemtree and custom for now.
+	 * @param {"itemtree" | "custom"} type 
+	 * @returns {ItemTreeColumnOptions[]}
+	 */
+	_getColumnsByType(type) {
+		type = type || "itemtree";
+		if (type === "itemtree") {
+			return ITEMTREE_COLUMNS.map(opt => Object.assign({}, opt));
+		}
+		if (type === "custom") {
+			return Object.values(this._customColumns).map(opt => Object.assign({}, opt));
+		}
+	}
+
+	/**
+	 * Check if column options is valid.
+	 * If the options is an array, all its children must be valid.
+	 * Otherwise, the validation fails.
+	 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options 
+	 * @returns {boolean} true if the option(s) are valid
+	 */
+	_validateColumnOption(options) {
+		/**
+		 * Validate column options.
+		 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options 
+		 * @param {(option: ItemTreeColumnOptions) => boolean} validator - A function that returns true if the option is valid
+		 * @returns {boolean} true if the option(s) are valid
+		 */
+		function validate(options, validator) {
+			if (Array.isArray(options)) {
+				return options.every(opt => validator(opt));
+			}
+			return validator(options);
+		}
+
+		// Check if the input option has duplicate dataKeys
+		const noInputDuplicates = !Array.isArray(options) || !options.find((opt, i, arr) => arr.findIndex(o => o.dataKey === opt.dataKey) !== i);
+		if (!noInputDuplicates) {
+			Zotero.warn(`ItemTree Column options have duplicate dataKey.`);
+		}
+		const requiredProperties = validate(options, (option) => {
+			const valid = option.dataKey && option.label;
+			if (!valid) {
+				Zotero.warn(`ItemTree Column option ${JSON.stringify(option)} must have dataKey and label.`);
+			}
+			return valid;
+		});
+		const noRegisteredDuplicates = validate(options, (option) => {
+			const valid = this.getColumns({ dataKey: option.dataKey }).length === 0;
+			if (!valid) {
+				Zotero.warn(`ItemTree Column option ${JSON.stringify(option)} with dataKey ${option.dataKey} already exists.`);
+			}
+			return valid;
+		});
+		return noInputDuplicates && requiredProperties && noRegisteredDuplicates;
+	}
+
+
+	/**
+	 * Add a new column or new columns.
+	 * If the options is an array, all its children must be valid.
+	 * Otherwise, no columns are added.
+	 * @param {ItemTreeColumnOptions | ItemTreeColumnOptions[]} options - An option or array of options to add
+	 * @returns {string | string[] | false} - The dataKey(s) of the added column(s) or false if no columns were added
+	 */
+	_addColumns(options) {
+		const isSingle = !Array.isArray(options);
+		if (isSingle) {
+			options = [options];
+		}
+		this._namespacedDataKey(options);
+		// If any check fails, return check results
+		if (!this._validateColumnOption(options)) {
+			return false;
+		}
+		for (const opt of options) {
+			this._customColumns[opt.dataKey] = Object.assign({}, opt, { custom: true });
```

Comment:
Referencing https://github.com/zotero/zotero/pull/3186/files#r1262808957

This is exactly where it's no longer clear without looking at `this._namespacedDataKeys()` whether we're registering columns with namespaced data keys.

Reply:
Explanation here: https://github.com/zotero/zotero/pull/3186#discussion_r1263253104